### PR TITLE
Do not warn about full backward hooks when only pre-hooks are registered (#189093)

### LINKED ISSUE #189093: Misleading full-backward-hook warning when only `register_full_backward_pre_hook` is used
## Issue

`nn.Module.register_full_backward_pre_hook` emits a warning worded as a
`register_full_backward_hook` warning when a module's forward inputs do not
require gradients, even though no full backward hook is registered.

The warning is:

```text
Full backward hook is firing when gradients are computed with respect to module outputs since no inputs require gradients. See https://docs.pytorch.org/docs/main/generated/torch.nn.Module.html#torch.nn.Module.register_full_backward_hook for more details.
```

This looks like a misfire or at least a misleading warning for
`register_full_backward_pre_hook`:

- the reproducer only calls `register_full_backward_pre_hook`
- it never calls `register_full_backward_hook`
- the warning points users at `register_full_backward_hook` docs
- full backward pre-hooks only receive `grad_output`, so firing from the output
  side when no inputs require grad is expected and useful

This is noisy for training-style modules whose batch input tensors usually do
not require gradients, while the module parameters do. In that case, a
full-backward pre-hook is still useful because it can prepare module state after
output gradients are available and before parameter gradients are computed.

## Minimal Repro

```python
import torch
from torch import nn


class ToyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.proj = nn.Linear(4, 4, bias=False)
        self.register_full_backward_pre_hook(lambda module, grad_output: None)

    def forward(self, x):
        return self.proj(x).relu()


model = ToyModule()
x = torch.randn(2, 4)  # does not require grad

model(x).sum().backward()
```

Pytest repro with additional controls:

```bash
python -m pytest -q repros/test_pytorch_full_backward_pre_hook_warning.py
```

The repro checks:

- one input that does not require grad: passes with the unexpected warning
- one input that requires grad: passes
- multiple inputs with no input requiring grad: passes with the unexpected warning
- multiple inputs where only one input requires grad: passes
- multiple outputs with no input requiring grad: passes with the unexpected warning
- multiple inputs/outputs where only one input requires grad: no warning is
  emitted and the test passes

## Observed Output

Validated locally with:

```text
torch=2.12.0a0+0291f960b6.nv26.04.48445190
device=cpu
```

The pytest repro passes all hook-call assertions, but it reports warnings for
the three cases where no input requires grad:

```text
repros/test_pytorch_full_backward_pre_hook_warning.py::test_single_input_no_grad
repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_no_grad
repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_outputs_no_input_grad
  UserWarning: Full backward hook is firing when gradients are computed with respect to module outputs since no inputs require gradients. See https://docs.pytorch.org/docs/main/generated/torch.nn.Module.html#torch.nn.Module.register_full_backward_hook for more details.
```

The controls pass without that warning:

```text
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_single_input_requires_grad
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_partially_require_grad
PASSED repros/test_pytorch_full_backward_pre_hook_warning.py::test_multiple_inputs_outputs_partially_require_grad
```

The pytest summary is:

```text
6 passed, 3 warnings
```

The observed mismatch is that:

- the repro only registers `register_full_backward_pre_hook`
- no `register_full_backward_hook` is registered
- pytest still reports the full-backward-hook warning for the no-input-grad cases

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by Functionhx**
## Issue

Fixes #189093

## Summary

BackwardHook fired the full-backward-hook warning whenever no module input
required grad, even when only a full backward pre-hook was registered
(pre-hooks only consume grad_output, so firing there is expected). The
warning and its fallback are now gated on a real full backward hook being
registered; pre-hook-only modules no longer get the spurious warning, and
behavior with a full backward hook is unchanged.

## Checklist

- [x] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by Functionhx** [comment]
The failing jobs all run the new hook tests under PYTORCH_TEST_WITH_DYNAMO=1. Full backward hooks are not reliably supported in Dynamo-wrapped tests, so both newly added hook tests are now gated with skipIfTorchDynamo. Their eager-mode coverage remains unchanged.
